core/imagedownloader.cpp

```diff
@@ -17,45 +17,50 @@ static QUrl cloudImageURL(const char *hash)
 ImageDownloader::ImageDownloader(struct picture *pic)
 {
 	picture = pic;
-	loadFromHash = false;
 }
 
 ImageDownloader::~ImageDownloader()
 {
 	picture_free(picture);
 }
 
-void ImageDownloader::load(bool fromHash){
-	QUrl url;
-	loadFromHash = fromHash;
-	if(fromHash)
-		url = cloudImageURL(picture->hash);
-	else
-		url = QUrl::fromUserInput(QString(picture->filename));
+void ImageDownloader::load(bool fromHash)
+{
+	if (fromHash && loadFromUrl(cloudImageURL(picture->hash)))
+		return;
+
+	// If loading from hash failed, try to load from filename
+	loadFromUrl(QUrl::fromUserInput(QString(picture->filename)));
+}
+
+bool ImageDownloader::loadFromUrl(const QUrl &url)
+{
+	bool success = false;
 	if (url.isValid()) {
 		QEventLoop loop;
 		QNetworkAccessManager manager;
 		QNetworkRequest request(url);
-		connect(&manager, SIGNAL(finished(QNetworkReply *)), this, SLOT(saveImage(QNetworkReply *)));
+		connect(&manager, &QNetworkAccessManager::finished, this,
+			[this,&success] (QNetworkReply *reply) { saveImage(reply, success); });
 		QNetworkReply *reply = manager.get(request);
```

Comment:
AFAIK, we don't have lambdas anywhere in the code. probably best to keep it that way.

Reply:
Actually, I stumbled over quite a few of them, so I thought it was OK.

Here, not using a lambda would mean adding a `success` member variable. Or do you have a different idea how we can get the download status back to the caller?

Reply:
maybe @dirkhh changed our policy on lambdas and they are now accepted?

Reply:
I still hate them. I'm aware of a few that I had to agree made code more readable - and it's typically in "connect()" situations. I hope I'll find time to review things (and do 4.7.7 - this will definitely not go in before that) tomorrow.

Reply:
Actually, stateful lambdas are one of the features I really like about C++. Just think about all the inane boiler-plate that you can avoid by using a simple lambda instead of a functor.